lib/src/collection/lru_map.dart

```diff
@@ -0,0 +1,286 @@
+  static const _DEFAULT_MAXIMUM_SIZE = 100;
+
+  final Map<K, _LinkedEntry<K, V>> _entries;
+
+  int _maximumSize;
+
+  _LinkedEntry<K, V> _head;
+  _LinkedEntry<K, V> _tail;
+
+  /**
+   * Create a new LinkedLruHashMap with a [maximumSize].
+   */
+  factory LinkedLruHashMap({int maximumSize}) =>
+      new LinkedLruHashMap._fromMap(new HashMap<K, _LinkedEntry<K, V>>(),
+          maximumSize: maximumSize);
+
+  LinkedLruHashMap._fromMap(
+      this._entries, {
+      int maximumSize})
+          // This pattern is used instead of a default value because we want to
+          // be able to respect null values coming in from MapCache.lru.
+          : _maximumSize = maximumSize != null ? maximumSize
+              : _DEFAULT_MAXIMUM_SIZE;
+
+  /**
+   * Iterates over [other], adding all key-value pairs to the map. The MRU/LRU
+   * order is based off of [other.forEach] implementation - the first item would
+   * be the LRU, and the last item, the MRU.
+   *
+   * Because [maximumSize] is still enforced, this means if [other.length] is
+   * > [maximumSize], not all of the items in [other] may exist at the end of the
+   * operation.
+   */
+  @override
+  void addAll(Map<K, V> other) => other.forEach((k, v) => this[k] = v);
+
+  @override
+  void clear() {
+    _entries.clear();
+    _head = _tail = null;
+  }
+
+  @override
+  bool containsKey(K key) => _entries.containsKey(key);
+
+  @override
+  bool containsValue(V value) => values.contains(value);
+
+  /**
+   * Applies [action] to each key-value pair of the map in order of MRU to LRU.
+   *
+   * Calling `action` must not add or remove keys from the map.
+   */
+  @override
+  void forEach(void action(K key, V value)) {
+    var head = _head;
+    while (head != null) {
+      action(head.key, head.value);
+      head = head.next;
+    }
+  }
+
+  @override
+  int get length => _entries.length;
+
+  @override
+  bool get isEmpty => _entries.isEmpty;
+
+  @override
+  bool get isNotEmpty => _entries.isNotEmpty;
+
+  /**
+   * The keys of [this] - in order of MRU to LRU.
+   *
+   * The returned iterable does *not* have efficient `length` or `contains`.
+   */
+  @override
+  Iterable<K> get keys => new _LinkedEntryIterator(_head).map((e) => e.key);
+
+  /**
+   * The values of [this] - in order of MRU to LRU.
+   *
+   * The returned iterable does *not* have efficient `length` or `contains`.
+   */
+  @override
+  Iterable<V> get values => new _LinkedEntryIterator(_head).map((e) => e.value);
+
+  @override
+  int get maximumSize => _maximumSize;
+
+  @override
+  void set maximumSize(int maximumSize) {
+    if (maximumSize == null) throw new ArgumentError.notNull('maximumSize');
+    while (length > maximumSize) {
+      _removeLru();
+    }
+    _maximumSize = maximumSize;
+  }
+
+  /**
+   * Look up the value of [key], or add a new value if it isn't there.
+   *
+   * Returns the value associated to [key], if there is one, and promotes the
+   * key-value pair to the MRU position.
```

Comment:
The second paragraph seems to repeat a lot of what the first one said. Can we combine the two?